// coc-lists/lib/windows.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const safe_1 = tslib_1.__importDefault(require("colors/safe"));
class Windows {
    constructor(nvim) {
        this.nvim = nvim;
        this.name = 'windows';
        this.description = 'windows list';
        this.defaultAction = 'jump';
        this.actions = [];
        this.actions.push({
            name: 'jump',
            execute: item => {
                if (Array.isArray(item))
                    return;
                nvim.call('win_gotoid', item.data.id, true);
            }
        });
    }
    async loadItems(_context) {
        let { nvim } = this;
        let wins = await nvim.call('getwininfo');
        let res = [];
        for (let win of wins) {
            let bufname = await nvim.call('bufname', win.bufnr);
            res.push({
                label: `${safe_1.default.yellow(win.tabnr.toString())}\t${safe_1.default.yellow(win.winnr.toString())}\t${bufname}`,
                data: { id: win.winid }
            });
        }
        return res;
    }
}
exports.default = Windows;
